File: day2.py
with open('./input2.txt') as f:
    
    txt = f.readlines()
    total_points = 0

    vals = {
        'A': 1,
        'B': 2,
        'C': 3,
        'X': 1,
        'Y': 2,
        'Z': 3,
    }

    for line in txt:
        line = line.strip('\n')
        points = 0
        points += vals[line[-1]]

        # if the value is same => draw
        if vals[line[-1]] == vals[line[0]]:
            points += 3
        
        # if difference > 1 => not adjecent => smaller number wins (rock beats scissors)
        elif abs(vals[line[-1]] - vals[line[0]]) == 2:
            if vals[line[0]] < vals[line[-1]]:
                points += 0
            else: points += 6

        # adjacent => greater beats smaller
        elif vals[line[-1]] > vals[line[0]]:
            points += 6

        total_points += points

    print(total_points)


# part 2

def custom_add(inp):
    if inp == 3: return 1
    return inp + 1

def custom_sub(inp):
    if inp == 1: return 3
    return inp - 1

with open('./input2.txt') as f:
    
    txt = f.readlines()
    total_points = 0

    vals = {
        'A': 1,
        'B': 2,
        'C': 3,
        'X': 0,
        'Y': 3,
        'Z': 6,
    }

    for line in txt:
        line = line.strip('\n')
        points = 0

        # add points for the outcome
        points += vals[line[-1]]

        # if need to lose
        if line[-1] == 'X':
            points += custom_sub(vals[line[0]])
        # if draw
        elif line[-1] == 'Y':
            points += vals[line[0]]
        elif line[-1] == 'Z':
            points += custom_add(vals[line[0]])

        total_points += points

    print(total_points)

File: day2.go
package main

import (
	"bufio"
	"fmt"
	//"math"
	"os"
)

func Day2() {
    file, err := os.Open("./input2.txt")
    if err != nil {
        panic(err)
    }
    defer file.Close()

    weapons := map[string]int{
        "A": 1,
        "B": 2,
        "C": 3,
        "X": 1,
        "Y": 2,
        "Z": 3,
    }
    points1 := 0
    points2 := 0

    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        round := scanner.Text()

        // Part 1
        // points1 += weapons[string(round[2])]
        // if weapons[string(round[0])] == weapons[string(round[2])] {
        //     points1 += 3
        //     continue
        // } 

        // if math.Abs(float64(weapons[string(round[0])] - weapons[string(round[2])])) > 1 {
        //     if weapons[string(round[0])] >= weapons[string(round[2])] {
        //         points1 += 6
        //     }
        // } else {
        //     if weapons[string(round[0])] <= weapons[string(round[2])] {
        //         points1 += 6
        //     }

        // }

        // Part 2
        if string(round[2]) == "X" { // need to lose
            points2 += 0 // losing
            points2 += wrappedSub(weapons[string(round[0])])
        } else if string(round[2]) == "Y" { // draw
            points2 += 3
            points2 += weapons[string(round[0])] // same weapon since we're drawing
        } else if string(round[2]) == "Z" { // win
            points2 += 6
            points2 += wrappedAdd(weapons[string(round[0])])
        }
    }

    fmt.Println(points1)
    fmt.Println(points2)
}

func wrappedAdd(start int) int {
    if start == 3 { return 1 }
    start += 1
    return start
}

func wrappedSub(start int) int {
    if start == 1 { return 3 }
    start -= 1
    return start
}

// Alternate way
type Weapon struct {
    points  int
    beats   *Weapon
    loses   *Weapon
}

type WeaponsLinkedList struct {
    head *Weapon
}

func Day2Alt() {
    file, err := os.Open("./input2.txt")
    if err != nil {
        panic(err)
    }
    defer file.Close()

    var rock Weapon
    var paper Weapon
    var scissor Weapon

    rock = Weapon{points: 1, beats: &scissor, loses: &paper}
    paper = Weapon{points: 2, beats: &rock, loses: &scissor}
    scissor = Weapon{points: 3, beats: &paper, loses: &rock}
    //weapons := WeaponsLinkedList{&rock}

    textToWeapons := map[string]Weapon{
        "A": rock,
        "B": paper,
        "C": scissor,
        "X": rock,
        "Y": paper,
        "Z": scissor,
    }

    points1 := 0
    scanner := bufio.NewScanner(file)
    for scanner.Scan() {
        round := scanner.Text()
        myWeapon := textToWeapons[string(round[2])]
        enemyWeapon := textToWeapons[string(round[0])]
        points1 += myWeapon.points

        if enemyWeapon == myWeapon { // draw
            points1 += 3
        } else {
            if myWeapon.beats.points == enemyWeapon.points { // I win
                points1 += 6
            } else { // I lose
                continue
            }
        }
    }

    fmt.Println("Points 1:", points1)
}
